src/vm/contexts.rs

```diff
@@ -39,9 +43,184 @@ pub struct LocalContext <'a> {
+                _ => {
+                    panic!("Tried to construct environment with bad sender {}", sender);
+                }
+            }
+        }
+        Environment {
+            global_context: global_context,
+            contract_context: contract_context,
+            call_stack: call_stack,
+            sender: sender
+        }
+    }
+
+    pub fn nest_with_sender <'c> (&'c mut self, sender: Value) -> Environment<'c, 'b> {
+        Environment::new(self.global_context,
+                         self.contract_context,
+                         self.call_stack,
+                         Some(sender))
+    }
+
+    pub fn eval_read_only(&mut self, contract_name: &str, program: &str) -> Result<Value> {
+        let parsed = parser::parse(program)?;
+        if parsed.len() < 1 {
+            return Err(Error::new(ErrType::ParseError("Expected a program of atleast length 1".to_string())))
+        }
+
+        let contract = self.global_context.database.get_contract(contract_name)?;
+        let mut nested_context = GlobalContext::begin_from(&mut self.global_context.database);
+        let result = {
+            let mut nested_env = Environment::new(&mut nested_context, &contract.contract_context, self.call_stack, self.sender.clone());
+            let local_context = LocalContext::new();
+            eval(&parsed[0], &mut nested_env, &local_context)
+        };
+        nested_context.database.roll_back();
+
+        result
+    }
+
+    pub fn execute_contract(&mut self, contract_name: &str, 
+                            tx_name: &str, args: &[SymbolicExpression]) -> Result<Value> {
+        let contract = self.global_context.database.get_contract(contract_name)?;
+        let mut nested_context = GlobalContext::begin_from(&mut self.global_context.database);
+        let result = {
+            let mut nested_env = Environment::new(&mut nested_context, &contract.contract_context, self.call_stack, self.sender.clone());
+            contract.execute_transaction(tx_name, args, &mut nested_env)
+        };
+
+        nested_context.handle_tx_result(result)
+    }
+
+    pub fn execute_function_as_transaction(&mut self, function: &DefinedFunction, args: &[Value]) -> Result<Value> {
+        let function = match function {
+            DefinedFunction::Private(_) => Err(Error::new(ErrType::NonPublicFunction(format!("{}", function.get_identifier())))),
+            DefinedFunction::Public(f) => Ok(f)
+        }?;
+
+        let mut nested_context = GlobalContext::begin_from(&mut self.global_context.database);
+
+        let result = {
+            let mut nested_env = Environment::new(&mut nested_context, self.contract_context, self.call_stack, self.sender.clone());
+
+            function.apply(args, &mut nested_env)
+        };
+
+        nested_context.handle_tx_result(result)
+    }
+
+    pub fn initialize_contract(&mut self, contract_name: &str, contract_content: &str) -> Result<()> {
```

Comment:
When I tried this out on a buggy contract, I noticed that the contract name gets taken even though the contract is buggy.  Can we make it so that only valid contracts can claim names?

Reply:
Yeah, that's a bug in `main`, actually, not here, since the `main.rs` code is handling the type check DB independently from the actual contract DB.

Reply:
So the bug you saw was because the type checker was failing subsequent initializations, not the VM.

Reply:
This behavior is now fixed -- `launch` now uses an outer-most savepoint to ensure that if the contract initialization fails in the interpreter, it will abort the type check database modification.